### PR TITLE
sha3 xor metric

### PR DESCRIPTION
**Pull request by subtly**

### CODE REVIEW

**Comment by chriseth** [comment]
Wouldn't it be better to directly use the sha3 of the address as NodeId, or at least store it inside Nodeid?

**Comment by subtly** [comment]
@chriseth Perhaps as an optimization. The output of the distance is already being stored.

File: libp2p/NodeTable.h [hunk 1]
```diff
@@ -138,7 +138,7 @@ class NodeTable: UDPSocketEvents, public std::enable_shared_from_this<NodeTable>
 	~NodeTable();
 
 	/// Returns distance based on xor metric two node ids. Used by NodeEntry and NodeTable.
-	static unsigned distance(NodeId const& _a, NodeId const& _b) { u512 d = _a ^ _b; unsigned ret; for (ret = 0; d >>= 1; ++ret) {}; return ret; }
+	static unsigned distance(NodeId const& _a, NodeId const& _b) { u256 d = sha3(_a) ^ sha3(_b); unsigned ret; for (ret = 0; d >>= 1; ++ret) {}; return ret; }
```
Comment (gavofyork) [inline libp2p/NodeTable.h]: is this called often? `sha3` is not an especially quick function to be calling...

**Comment by fjl** [comment]
For cross-reference, the go-ethereum PR is here: ethereum/go-ethereum#791.
The Go PR has a protocol change which makes findnode target hash-sized instead of public key sized.
Will submit a devp2p PR for that.

**Comment by fjl** [comment]
https://github.com/ethereum/devp2p/pull/25

File: libp2p/NodeTable.h (hunk 1 above)
Comment (subtly) [inline libp2p/NodeTable.h]: Not often. It's called by the constructor of NodeEntry and the result is stored in a property.

**Pull request merged**